core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala

```diff
@@ -211,51 +223,43 @@ class PaigesScalacheckTest extends FunSuite {
     }
   }
 
-  test("the left and right side of a union are the same after flattening") {
+  test("Union invariant: `a` consists of Text, Empty, Concat") {
     import Doc._
-    def okay(d: Doc): Boolean = d match {
-      case Empty | Text(_) | Line(_) => true
-      case Concat(a, b) => okay(a) && okay(b)
-      case Nest(j, d) => okay(d)
-      case Align(d) => okay(d)
-      case LazyDoc(d) => okay(d.evaluated)
-      case Union(a, b) =>
-        (a.flatten === b.flatten) && okay(a) && okay(b)
+    def ok(d: Doc): Boolean = d match {
+      case Concat(a, b) => ok(a) && ok(b)
+      case Text(_) | Empty => true
+      case _ => false
     }
+    forAll { d: Union => ok(d.a) }
+  }
 
-    forAll { (a: Doc) => assert(okay(a)) }
+  test("Union invariant: `a.flatten == a`") {
+    forAll { d: Union => d.a.flatten === d.a }
   }
 
-  test("the left side of a union has a next line as long or longer than the right") {
+  test("Union invariant: `a == b.flatten`") {
+    forAll { d: Union => d.a === d.b.flatten }
+  }
+
+  test("Union invariant: `a != b`") {
+    forAll { d: Union => !(d.a === d.b) }
+  }
+
+  test("Union invariant: `a` has right-associated `Concat` nodes") {
     import Doc._
+    def ok(d: Doc): Boolean = d match {
+      case Concat(Concat(_, _), _) => false
+      case Empty | Text(_) => true
+      case Concat(a, b) => ok(a) && ok(b)
+      case _ => throw new RuntimeException(s"Invariant failure: ${d}")
+    }
+    forAll { d: Union => ok(d.a) }
```

Comment:
`assert(ok(d.a))` is needed since scalatest sees that an assert was done.

Reply:
Thank you.  I'll fix all of these.  I tried adding WartRemover's `NonUnitStatements` but ScalaTest has so many that it was impossible to turn that on.  I see there's a way to extend a different trait in ScalaTest that makes ScalaCheck tests work like I imagined they would, but that would be a lot of churn.